### ISSUE TITLE
Unable to load model at full context length

### ISSUE THREAD

**Issue by abhinavkulkarni** [huggingface/text-generation-inference#778]
Hi,

I am trying to load `AutoGPTQ` quantized models with TGI. I know that TGI has its own implementation of GPTQ algorithm, but wanted to try out models outside of TGI ecosystem such `AutoGPTQ`, `AWQ`, etc.

With usual HuggingFace transformers API, I am able to load the following 7B Llama 2 `AutoGPTQ` model and infer at full 4096 context length with `model.generate` on my 12GB RTX 3060 card. In fact, the model uses less than 9GB of VRAM at full 4096 context length. So, I wanted to load the model at full 4096 context length in TGI.

Firstly, I overrode the `get_model` method to always return `CausalLM` if `--trust-remote-code` flag is passed.

```python
def get_model(
    model_id: str,
    revision: Optional[str],
    sharded: bool,
    quantize: Optional[str],
    dtype: Optional[str],
    trust_remote_code: bool,
) -> Model:
    if dtype is None:
        dtype = torch.float16
    elif dtype == "float16":
        dtype = torch.float16
    elif dtype == "bfloat16":
        dtype = torch.bfloat16
    else:
        raise RuntimeError(f"Unknown dtype {dtype}")

    if trust_remote_code:
        return CausalLM(
            model_id,
            revision,
            quantize=quantize,
            dtype=dtype,
            trust_remote_code=trust_remote_code,
        )

    raise ValueError(f"Unsupported model type {model_type}")
```

Secondly, I installed `AutoGPTQ` (`pip install auto_gptq`) and overrode the `__init__` method of `CausalLM` to load [7B Llama2 GPTQ model](https://huggingface.co/TheBloke/Llama-2-7b-Chat-GPTQ) in `CausalLM` as follows (as per instructions in the model card):

```python
class CausalLM(Model):
    def __init__(
        self,
        model_id: str,
        revision: Optional[str] = None,
        quantize: Optional[str] = None,
        dtype: Optional[torch.dtype] = None,
        trust_remote_code: bool = False,
    ):
        if torch.cuda.is_available():
            device = torch.device("cuda")
            dtype = torch.float16 if dtype is None else dtype
        else:
            if quantize:
                raise ValueError("quantization is not available on CPU")

            device = torch.device("cpu")
            dtype = torch.float32

        tokenizer = AutoTokenizer.from_pretrained(
            model_id,
            revision=revision,
            padding_side="left",
            truncation_side="left",
            trust_remote_code=trust_remote_code,
        )
        from auto_gptq import AutoGPTQForCausalLM, BaseQuantizeConfig
        use_triton = False
        model = AutoGPTQForCausalLM.from_quantized(model_id,
                use_safetensors=True,
                device="cuda:0",
                use_triton=use_triton,
                quantize_config=None)
        model.config.use_cache = True
        model = model.eval()

        if torch.cuda.is_available() and torch.cuda.device_count() == 1:
            model = model.cuda()

        if tokenizer.pad_token_id is None:
            if model.config.pad_token_id is not None:
                tokenizer.pad_token_id = model.config.pad_token_id
            elif model.config.eos_token_id is not None:
                tokenizer.pad_token_id = model.config.eos_token_id
            elif tokenizer.eos_token_id is not None:
                tokenizer.pad_token_id = tokenizer.eos_token_id
            else:
                tokenizer.add_special_tokens({"pad_token": "[PAD]"})

        super(CausalLM, self).__init__(
            model=model,
            tokenizer=tokenizer,
            requires_padding=True,
            dtype=dtype,
            device=device,
        )
```

With the following command, I get an OOM error.

```
text-generation-launcher \
--huggingface-hub-cache ~/.cache/huggingface/hub/ \
--model-id TheBloke/Llama-2-13b-Chat-GPTQ \
--trust-remote-code --port 8080 \
--max-input-length 4095 --max-total-tokens 4096 --max-batch-prefill-tokens 4096 --max-batch-total-tokens 4096
```

Thanks!

**Comment by abhinavkulkarni** [comment]
I get a similar OOM error if I try to load [AWQ quantized 7B Llama 2 model](https://huggingface.co/abhinavkulkarni/meta-llama-Llama-2-7b-chat-hf-w4-g128-awq). Again, it is possible to load this AWQ model at full 4096 context length on a 12GB of VRAM rather comfortably.

**Comment by abhinavkulkarni** [comment]
Hi,

It would be good to have some guidance on `warmup` method of `client.rs`. For cmd options `--max-input-length 4095 --max-total-tokens 4096 --max-batch-prefill-tokens 4096 --max-batch-total-tokens 4096`, it should send a batch of two inputs of length 4095 and 1 to the server. However, logging the `batch.input_ids` in `Warmup` method of `server.py`, I see two identical batches of length `4095`. 

Basically, given the max context size of `4096`, I am trying to maximize the `max_input_tokens`, but the warmup is failing because it sends a batch of `2x4095`, thus causing OOM error.

Let me know if there's a way out of this.

**Comment by abhinavkulkarni** [comment]
So, I hardcoded the `warmup` method of `client.rs` to return `Ok(Some(4096))` to mitigate the warmup OOM issue.

But then, because this is a custom model (even though it's based on `LlamaForCausalLM` architecture, TGI is not able to find the fast tokenizer. With `--max-input-length 4095 --max-total-tokens 4096 --max-batch-prefill-tokens 4096 --max-batch-total-tokens 4096`, TGI pads the input to `4095` however small it may, leaving no token for output.

**Comment by Narsil** [comment]
Try removing all the custom code you have written and simply load your model with 

`--quantize gptq` 

As long as your model has this file: https://huggingface.co/TheBloke/Llama-2-13B-chat-GPTQ/blob/main/quantize_config.json it should load properly and run as expected.

Also no need to disable warmup, just use lower requirements  (`--max-concurrent-requests = ` if you really don´t  care about concurrent users)

**Comment by 0x1997** [comment]
Same here. Can't load LLaMA-2 70B GPTQ model with only '--quantize gptq' flag.
I'm using a A100 GPU and the model size is about 35G.

```
Method Warmup encountered an error.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/models/flash_causal_lm.py", line 727, in warmup
    _, batch = self.generate_token(batch)
  File "/opt/conda/lib/python3.9/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/models/flash_causal_lm.py", line 825, in generate_token
    raise e
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/models/flash_causal_lm.py", line 813, in generate_token
    out = self.forward(
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/models/flash_causal_lm.py", line 789, in forward
    return self.model.forward(
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/models/custom_modeling/flash_llama_modeling.py", line 478, in forward
    hidden_states = self.model(
  File "/opt/conda/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/models/custom_modeling/flash_llama_modeling.py", line 437, in forward
    hidden_states, residual = layer(
  File "/opt/conda/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/models/custom_modeling/flash_llama_modeling.py", line 381, in forward
    mlp_output = self.mlp(normed_attn_res_output)
  File "/opt/conda/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/models/custom_modeling/flash_llama_modeling.py", line 327, in forward
    return self.down_proj(self.act(gate_up_states[:, 0]) * gate_up_states[:, 1])
RuntimeError: CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/conda/bin/text-generation-server", line 8, in <module>
    sys.exit(app())
  File "/opt/conda/lib/python3.9/site-packages/typer/main.py", line 311, in __call__
    return get_command(self)(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/typer/core.py", line 778, in main
    return _main(
  File "/opt/conda/lib/python3.9/site-packages/typer/core.py", line 216, in _main
    rv = self.invoke(ctx)
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/typer/main.py", line 683, in wrapper
    return callback(**use_params)  # type: ignore
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/cli.py", line 81, in serve
    server.serve(
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/server.py", line 184, in serve
    asyncio.run(
  File "/opt/conda/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/conda/lib/python3.9/asyncio/base_events.py", line 634, in run_until_complete
    self.run_forever()
  File "/opt/conda/lib/python3.9/asyncio/base_events.py", line 601, in run_forever
    self._run_once()
  File "/opt/conda/lib/python3.9/asyncio/base_events.py", line 1905, in _run_once
    handle._run()
  File "/opt/conda/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/opt/conda/lib/python3.9/site-packages/grpc_interceptor/server.py", line 159, in invoke_intercept_method
    return await self.intercept(
> File "/opt/conda/lib/python3.9/site-packages/text_generation_server/interceptor.py", line 21, in intercept
    return await response
  File "/opt/conda/lib/python3.9/site-packages/opentelemetry/instrumentation/grpc/_aio_server.py", line 82, in _unary_interceptor
    raise error
  File "/opt/conda/lib/python3.9/site-packages/opentelemetry/instrumentation/grpc/_aio_server.py", line 73, in _unary_interceptor
    return await behavior(request_or_iterator, context)
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/server.py", line 60, in Warmup
    max_supported_total_tokens = self.model.warmup(batch)
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/models/flash_causal_lm.py", line 729, in warmup
    raise RuntimeError(
RuntimeError: Not enough memory to handle 4096 prefill tokens. You need to decrease `--max-batch-prefill-tokens`
```

As the model context length is 4096 tokens, I'd like to set `--max_total_tokens 4096 --max-input-length 4000` to allow longer inputs. So I cannot really decrease `--max-batch-prefill-tokens` much.

**Comment by Narsil** [comment]
Try decreasing `--max-concurrent-requests` instead then.

Can you show `nvidia-smi` after load and decreasing `--max-batch-prefill-tokens` (just to see).
Seems odd that it would take that much memory indeed.

**Comment by Narsil** [comment]
Something must be wrong in your environment...

I just loaded 70B GPTQ on A100(80GB) and I'm getting
```
Setting max batch total tokens to 137648
```

Is it possible you have another process already on the GPU ?

**Comment by 0x1997** [comment]
@Narsil Setting `--max-concurrent-requests 1` has no effect.
With this flag, the max vram usage before crash is about 39GB.
I have [nvitop](https://github.com/XuehaiPan/nvitop) running to monitor the loading process and I can see there's no other processes on the GPU.

**Comment by Narsil** [comment]
Are you sure you are using 80GB then ? Could be you're only using/detecting 40GB (which would explain the error).

Dart in the dark which might be useful: https://github.com/huggingface/text-generation-inference/issues/714#issuecomment-1655207099

Can you provide more details about those A100 ? Which cloud, which setup, cuda+nvidia driver version maybe ?

**Comment by Narsil** [comment]
Check on different hardware it really seems to be that your environment is not correctly set up and that you're seeing only 40GB of VRAM

**Comment by 0x1997** [comment]
The card works fine for training and inference and I've seen vram usage up to near 80GB. The required nvidia services are also running.
The problem disappeared when loading with `--sharded true --num-shard 2`.
I'll test it on another machine next week.

**Comment by OlivierDehaene** [comment]
Closing for inactivity. Feel free to open an issue with the problem persists.

**Comment by sauravm8** [comment]
Any tips if I am facing the same error when loading the model with AutoModelForCausalLM.pre_trained()?
